lib/quantization/src/encoded_vectors_binary.rs

```diff
@@ -212,17 +314,79 @@ impl<TBitsStoreType: BitsStoreType, TStorage: EncodedStorage>
                 encoded_vector[i / bits_count] |= one << (i % bits_count);
             }
         }
+    }
 
-        EncodedBinVector { encoded_vector }
+    fn encode_two_bits_vector(
+        vector: &[f32],
+        encoded_vector: &mut [TBitsStoreType],
+        standard_deviations: &[f32],
+        means: &[f32],
+    ) {
+        let bits_count = u8::BITS as usize * std::mem::size_of::<TBitsStoreType>();
+        let one = TBitsStoreType::one();
+        for (i, &v) in vector.iter().enumerate() {
+            let mean = means[i];
+            let sd = standard_deviations[i];
+            let ranges = 3;
+            let v_z = (v - mean) / sd;
+            let index = (v_z + 2.0) / (4.0 / ranges as f32);
+
+            if index >= 1.0 {
+                let count_ones = (index.floor() as usize).min(2);
+                if count_ones > 1 {
+                    encoded_vector[i / bits_count] |= one << (i % bits_count);
+                }
+                if count_ones > 0 {
+                    let j = vector.len() + i;
+                    encoded_vector[j / bits_count] |= one << (j % bits_count);
+                }
+            }
+        }
     }
 
-    pub fn get_quantized_vector_size_from_params(vector_parameters: &VectorParameters) -> usize {
-        TBitsStoreType::get_storage_size(vector_parameters.dim)
+    fn encode_one_and_half_bits_vector(
+        vector: &[f32],
+        encoded_vector: &mut [TBitsStoreType],
+        standard_deviations: &[f32],
+        means: &[f32],
+    ) {
+        let bits_count = u8::BITS as usize * std::mem::size_of::<TBitsStoreType>();
+        let one = TBitsStoreType::one();
+        for (i, &v) in vector.iter().enumerate() {
+            let mean = means[i];
+            let sd = standard_deviations[i];
+            let ranges = 3;
+            let v_z = (v - mean) / sd;
+            let index = (v_z + 2.0) / (4.0 / ranges as f32);
+
+            if index >= 1.0 {
+                let count_ones = (index.floor() as usize).min(2);
+                if count_ones > 1 {
+                    encoded_vector[i / bits_count] |= one << (i % bits_count);
+                }
+                if count_ones > 0 {
+                    let j = vector.len() + i / 2;
+                    encoded_vector[j / bits_count] |= one << (j % bits_count);
+                }
+            }
+        }
```

Comment:
Requires commend with examples

Reply:
Done, I covered all logic by comments, see `fn encode_two_bits_value(value: f32, mean: f32, sd: f32)`. For 1.5bit quantization I added a motivation and examples how do we unite two 2bit values into 3bits